### PR TITLE
Fix flaky SPSA shot vector tests and add `num_directions` as supported `gradient_kwarg`

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
#3366 added the gradient transform `spsa_grad`
#3440 added QNode registration  of that transform
#3724 modified the tolerance in the transform tests (for the new return type system)

**Description of the Change:**
This PR fixes the (NumPy) seed of all SPSA related tests to make sure the tests are no longer flaky.

Also makes three more test files lint-compatible.

**Benefits:**
SPSA tests do not fail randomly on other CI runs.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**

### CODE REVIEW

File: tests/gradients/test_spsa_gradient.py [hunk 1]
```diff
@@ -30,17 +31,22 @@ def coordinate_sampler(indices, num_params, idx, seed=None):
     to the index ``indices[idx]``. This is a sequential coordinate sampler
     that allows to exactly reproduce derivatives, instead of using SPSA in the
     intended way."""
+    # pylint: disable=unused-argument
```
Comment (AlbertMitjans) [inline tests/gradients/test_spsa_gradient.py:34]: Why do we need this?

File: tests/gradients/test_spsa_gradient.py [hunk 2]
```diff
@@ -295,13 +307,14 @@ def test_y0_provided(self):
...
         n = 9
-        tapes, fn = spsa_grad(tape, strategy="forward", approx_order=1, num_directions=n, f0=f0)
+        tapes, _ = spsa_grad(tape, strategy="forward", approx_order=1, num_directions=n, f0=f0)
         # one tape per direction, the unshifted one already was evaluated above
         assert len(tapes) == n
 
     def test_independent_parameters(self):
         """Test the case where expectation values are independent of some parameters. For those
         parameters, the gradient should be evaluated to zero without executing the device."""
+        np.random.seed(42)
```
Comment (AlbertMitjans) [inline tests/gradients/test_spsa_gradient.py:317]: Where is the `np.random.seed` used? We were discussing not a long time ago that it is preferred to use a numpy [Random Generator](https://numpy.org/doc/stable/reference/random/generator.html) instead, to avoid having unwanted behaviours when the user fixes the random seed.

This is however outside the scope of this PR

File: tests/gradients/test_spsa_gradient.py [hunk 3]
```diff
@@ -443,10 +463,11 @@ def reference_qnode(x):
         assert np.isclose(qnode(par).item().val, reference_qnode(par))
         assert np.isclose(qml.jacobian(qnode)(par).item().val, qml.jacobian(reference_qnode)(par))
 
-    @pytest.mark.parametrize("num_directions, tol", [(100, 0.1), (500, 5e-3)])
+    @pytest.mark.parametrize("num_directions, tol", [(1, 0.1), (1, 5e-3)])
```
Comment (AlbertMitjans) [inline tests/gradients/test_spsa_gradient.py:466]: Why this change?

File: tests/returntypes/gradients/test_spsa_gradient_new.py [hunk 4]
```diff
@@ -658,7 +667,10 @@ def test_single_expectation_value_with_argnum_one(self, approx_order, strategy,
         assert np.allclose(res, expected, atol=tol, rtol=0)
 
     def test_multiple_expectation_value_with_argnum_one(
-        self, approx_order, strategy, validate, tol
+        self,
+        approx_order,
+        strategy,
+        validate,
```
Comment (AlbertMitjans) [inline tests/returntypes/gradients/test_spsa_gradient_new.py:673]: remove comma

**Review by AlbertMitjans — COMMENTED** [review]
Thank you for linting more test files! ❤️ 

I'll approve after the following comments are addressed 👍

File: tests/gradients/test_spsa_gradient.py (hunk 2 above)
Comment (dwierichs) [reply tests/gradients/test_spsa_gradient.py:317]: SPSA is an inherently stochastic method. It uses random samples to estimate the gradient.

I think the code of SPSA should then indeed be upgraded to use a random generator?

File: tests/gradients/test_spsa_gradient.py (hunk 3 above)
Comment (dwierichs) [reply tests/gradients/test_spsa_gradient.py:466]: Ah, sorry. I tested whether this test really tests what it says. And it doesn't: For a single parameter, there is no stochasticity in SPSA. Will modify so that this test does what is says.

Comment (dwierichs) [reply tests/gradients/test_spsa_gradient.py]: now testing two parameters, so that there is a convergence to test for.
Now requires 10 seconds on my laptop for the many-shots test, do we want to keep it?

File: tests/gradients/test_spsa_gradient.py (hunk 1 above)
Comment (AlbertMitjans) [reply tests/gradients/test_spsa_gradient.py:34]: Ah! In PL the `# pylint: disable` message is usually written **above** the function (hence my confusion).

Comment (dwierichs) [reply tests/gradients/test_spsa_gradient.py:34]: Oh oops! Shall I move it to the line exactly above the `def ` statement? I wasn't aware sorry :D

File: tests/gradients/test_spsa_gradient.py (hunk 3 above)
Comment (AlbertMitjans) [reply tests/gradients/test_spsa_gradient.py]: Sorry, which is the "many-shots" test?

File: tests/gradients/test_spsa_gradient.py (hunk 1 above)
Comment (AlbertMitjans) [reply tests/gradients/test_spsa_gradient.py:34]: It's not a big deal 😄 

You can leave it like this

File: tests/gradients/test_spsa_gradient.py (hunk 3 above)
Comment (dwierichs) [reply tests/gradients/test_spsa_gradient.py]: Ah, oops. many directions! 
The second parametrized test here (which now uses 10k directions)

Comment (AlbertMitjans) [reply tests/gradients/test_spsa_gradient.py]: Ah! I see... could we reduce the number of dimensions? Or just test it with 100 dimensions and remove the second parametrization

Comment (dwierichs) [reply tests/gradients/test_spsa_gradient.py]: I increased the tolernaces a bit and this allowed to do 100 and 1000 directions, which both is fast enough :)

**Review by AlbertMitjans — APPROVED** [review]
Good to go!! 🚀
